Refactor to support multiples environments

adalessa opened the pull request:
The idea of this change is to support multiple environment like docker compose, or more configurations for environments more flexible.

dorkster100 commented:
Also this does not seem to be checking for the laravel project, so if used in any other project i get constant notifications that the app cannot be initialized

adalessa commented:
@dorkster100 should be fixed. if the environment can't be initialize because the artisan file is not present will silently end the process. That message was intended when was a laravel proyect.

dorkster100 commented:
> @dorkster100 should be fixed. if the environment can't be initialize because the artisan file is not present will silently end the process. That message was intended when was a laravel proyect.

it still says that it can't initialize the environment, it is not able to get through the first .env check, i think that at the time when the plugin is trying to do DogenvGet, the .env file is still not loaded.

adalessa commented:
Have you set the variable in the environment? I fixed because should not stop there, The problem was the value I was expecting nill but it was returned as empty string. I have added a check for that.

dorkster100 commented:
> Have you set the variable in the environment? I fixed because should not stop there, The problem was the value I was expecting nill but it was returned as empty string. I have added a check for that.

I added some other suggested fixes, after fixing them locally everything seems to work fine so far.

dorkster100 commented:
@adalessa it is still not working for me, did you see my last comments?

adalessa commented:
@dorkster100  I miss understand you, when you meant that is was working fine so far. Got the impression that it was done. Can you open an issue with the error details of what is failing in detail

dorkster100 commented:
> @dorkster100 I miss understand you, when you meant that is was working fine so far. Got the impression that it was done. Can you open an issue with the error details of what is failing in detail

@adalessa There are several issues, it would be a bit difficult to separate them into different issues, have you seen my comments in the code in this pull request? They basically describe the issues with relation to specific block of code, as I have already done debugging to pinpoint the issue.

I can make a short summary:

* I am not able to pass container name, because the environments are setup before options are merged and nothing is passed as parameter. The best solution is probably to use some .env variable to be able to have different container names per project.
* In docker compose.lua unpack does not work as expected.
* route.lua contains json == "" is not enough, you need to add json == {} check as well.

adalessa commented:
@dorkster100 there are no comments on the code. Maybe you started the review and do not finished it ?
I will try to reproduce these and fix them.

dorkster100 commented:
> @dorkster100 there are no comments on the code. Maybe you started the review and do not finished it ? I will try to reproduce these and fix them.

@adalessa sorry, was not aware it requires me to submit my comments

dorkster100 commented on lua/laravel/environment/resolver.lua:
```diff
@@ -0,0 +1,70 @@
+  return function(environments)
+    -- TODO: base on the arguments resolve which environment should be use or nil
+    -- if no one matches the configuration
+    local env_name = env "NVIM_LARAVEL_ENV"
```
I found out in the end that dependency is this: https://github.com/tpope/vim-dotenv
Should probably add this to the readme

adalessa replied:
agree I have added it. This is usefull to read the from the .env because they are not loaded into neovim or the environment before.

adalessa commented on lua/laravel/environment/docker_compose.lua:
```diff
@@ -0,0 +1,29 @@
+M.setup = function(opts)
+  opts = opts or {}
+
+  local container = opts.container_name or "app"
```
I have added it as first option. If not present will use passed variable. if not app.

adalessa commented on lua/laravel/user-commands/sail.lua:
```diff
@@ -1,9 +1,11 @@
+local application = require "laravel.application"
 
 local commands = {
   up = function()
```
Yes I went with a single point if not each command that you may want to run will need to check for it. If you add customs executables you can use the same run command.

adalessa commented on lua/laravel/environment/docker_compose.lua:
```diff
@@ -0,0 +1,29 @@
+  local cmd = opts.cmd or { "docker", "compose", "exec", "-it", container }
+
+  local function get_cmd()
+    return unpack(cmd)
```
I was not aware of this. Fixed with your implementation

adalessa commented on lua/laravel/_config.lua:
```diff
@@ -31,6 +31,14 @@ local config = {
+  environment = {
+    resolver = require "laravel.environment.resolver"(true, true, nil),
+    environments = {
+      ["local"] = require("laravel.environment.native").setup(),
```
My idea with this is that you can easily overwrite the configuration calling the function. This makes if I add from the suported a new field you don't need to modify your local setup since the function is the one being executed.
Not in this file you will not pass variables but in your call to the `require("laravel").setup()` there you can set it.
